### ISSUE TITLE
Some problems on MIPS platform.

### ISSUE THREAD

**Issue by tobbe2007** [wkhtmltopdf/wkhtmltopdf#3621]
Hi, I have transplanted wkthmltox to Mips platform. It can run but have some problems while I convert html to pdf. Follow is the detail.
(note: I never learn C++)
Step 1: I spent server days to transplant it to Mips platforms and it worked.
Step 2: I upload it to embed-Mips device, run the command "./wkhtmltopdf test.html test.pdf".  Here is the error msg:
      "QFontDatabase::loadFromCache: Could not open font database cache!""
Step 3: I modified the source code: qt/src/gui/text/qfontdatabase_qws.cpp  at line 164:
`
163     #ifdef Q_WS_QWS
164     // const bool weAreTheServer = QWSServer::instance(); // this is the orignal code
165     const bool weAreTheServer = true; // this is my modification
166 #else
167     const bool weAreTheServer = true; // assume single-process
168 #endif
`
       Now, It can convert the pure English html files to pdf, but it core dumped if there have some Chinese chars in the html files.
      I think this is caused by my violent modification.
      May be it can worked well if "QWSServer::instance()"  is true when we start the Qt, but I don't understand the C++ code. Please help me.
      Note: Some boy said it should add param '-aws' while you start the Qt.
      Thanks.

**Comment by tobbe2007** [comment]
@NiR- Please help me. Thanks!

**Comment by NiR-** [comment]
Hello,

I know pretty much nothing about MIPS architecture and have no experience with Qt for embedded linux. Anyway, here are my thoughts. It took me a while and there might be more noise than signal but I hope it'll help you :)

First, could you provide more details about following points please?
* Your OS, and its version
* The version of wkhtmltopdf
* Do you use the patched QT?
* Which font packages did you installed?

Just out of curiosity (and it might help other people too), would you mind to share more details about how you build wkhtmltopdf for your embedded hardware?

#### TL;DR

You should try to:
* Revert your previous change
* Change [the third argument of QApplication's constructor in wkhtmltopdf.cc](https://github.com/wkhtmltopdf/wkhtmltopdf/blob/333521a51bf53ba632f3281a73fd14091c3ab250/src/pdf/wkhtmltopdf.cc#L189), to enforce `QApplication::GuiServer`.

#### Long explanation

> Step 3: I modified the source code: qt/src/gui/text/qfontdatabase_qws.cpp at line 164:
163 #ifdef Q_WS_QWS 164 // const bool weAreTheServer = QWSServer::instance(); // this is the orignal code 165 const bool weAreTheServer = true; // this is my modification 166 #else 167 const bool weAreTheServer = true; // assume single-process 168 #endif

IMHO, this isn't the right fix. [From the documentation](http://doc.qt.io/qt-4.8/qwsserver.html#instance), it appears that, `QWSServer::instance()`, when casted to boolean, would be true only if the current process is the actual rendering server.

> "QFontDatabase::loadFromCache: Could not open font database cache!""

I think the right way to make wkhtmltopdf working is to fix this error. To find out what file is missing there're basically two ways: either by reading the code, or by using some debug techniques.

So:
* the error message you provided comes from [here](https://github.com/wkhtmltopdf/qt/blob/8dae19a82231e87366d81c683bafcb025aea9c1d/src/gui/text/qfontdatabase_qws.cpp#L190), because the file `binaryDb` does not exist.
* `binaryDb` (the missing file) is defined [here](https://github.com/wkhtmltopdf/qt/blob/8dae19a82231e87366d81c683bafcb025aea9c1d/src/gui/text/qfontdatabase_qws.cpp#L171). And takes as value the concatenation of `qws_fontCacheDir()` with `/fontdb`.
* `qws_fontCacheDir()` is defined [here](https://github.com/wkhtmltopdf/qt/blob/8dae19a82231e87366d81c683bafcb025aea9c1d/src/gui/text/qfontengine_qpf.cpp#L231). And takes as value the concatenation of `qws_dataDir()` with `/fonts/`. And if that directory doesn't exist it returns the path to a temporary directory.
* `qws_dataDir()` is defined [here](https://github.com/wkhtmltopdf/qt/blob/8dae19a82231e87366d81c683bafcb025aea9c1d/src/gui/kernel/qapplication_qws.cpp#L181). It is based on the value of the `QT_VFB_DATADIR` macro and uses the id of the current display. It checks if the directory exists and if it has the right permissions.
* `QT_VFB_DATADIR` is defined [here](https://github.com/wkhtmltopdf/qt/blob/8dae19a82231e87366d81c683bafcb025aea9c1d/src/gui/embedded/qvfbhdr.h#L74). It concatenates `/tmp` with `/qtembedded-<DISPLAYID>`.

We need to step back and see what `QFontDatabasePrivate::loadFromCache()` is actually doing:
* First, note that [this method is declared only if `QT_FONTS_ARE_RESOURCES` isn't defined](https://github.com/wkhtmltopdf/qt/blob/0a2f2382541424726168804be2c90b91381608c6/src/gui/text/qfontdatabase_qws.cpp#L160). See below, it might be interesting for you.
* At the beginning of the method, [there're some early returns](https://github.com/wkhtmltopdf/qt/blob/8dae19a82231e87366d81c683bafcb025aea9c1d/src/gui/text/qfontdatabase_qws.cpp#L173) when it's called by the server process and either the path at `fontPath` or the path at `fontPath` + "/fontdir" are newer.
* Basically, it loads some font families from a cache file and put them into memory, unless they're stale.

What's interesting about that function is:
* It's called only by `initializeDb()` function, which in turn is called in four places. But the function takes care to be executed only once.
* [If the font cache is stale or doesn't exist](https://github.com/wkhtmltopdf/qt/blob/0a2f2382541424726168804be2c90b91381608c6/src/gui/text/qfontdatabase_qws.cpp#L308), [it loads fonts from the `fontDirFile`](https://github.com/wkhtmltopdf/qt/blob/0a2f2382541424726168804be2c90b91381608c6/src/gui/text/qfontdatabase_qws.cpp#L320).
* `fontDirFile` is the concatenation of `qwsFontPath()` with "/fontdir". And `qwsFontPath()` return either the env var QT_QWS_FONTDIR (when defined at runtime) or `/libs/fonts`.
* It uses `QFONTDATABASE_DEBUG` to toggle some debug mesages about loaded fonts. It might be interesting to enable it to have better insight of which fonts are actually loadedx.

To conclude:
* `binaryDb` is either `/tmp/qtembedded<DISPLAYID>/fonts/fontdb` or [`$TMPDIR/fontdb`](https://github.com/wkhtmltopdf/qt/blob/8dae19a82231e87366d81c683bafcb025aea9c1d/src/corelib/io/qdir.cpp#L1911)
* It's used only if current process does not act as a rendering server or when the cache files are newer.
* Thanks to `fontPath` argument (`fontDirFile` in `initializeDb()`) and `QT_QWS_FONTDIR` env var, we might be able to initialize the font cache used.

So what's strange is that you needed to force the value of `weAreTheServer`. As mentioned by [this page](http://doc.qt.io/qt-4.8/qt-embedded-architecture.html) and [this page](http://doc.qt.io/qt-4.8/qt-embedded-running.html), embedded applications needs a server or should be their own server. It looks like you could do this by adding  `-qws` to any Qt application (check [here](http://doc.qt.io/qt-4.8/qt-embedded-running.html) and [here](http://doc.qt.io/qt-4.8/qt-embedded-displaymanagement.html) for more details). [But that's equivalent to passing `QApplication::GuiServer` as the third argument of QApplication](http://doc.qt.io/qt-4.8/qapplication.html#QApplication-3).

And here is the problem:
* `QApplication` constructor has several signatures, and the one used [in wkhtmltopdf.cc](https://github.com/wkhtmltopdf/wkhtmltopdf/blob/333521a51bf53ba632f3281a73fd14091c3ab250/src/pdf/wkhtmltopdf.cc#L189) has 3 arguments too, but not the same: [`QApplication::QApplication(int & argc, char ** argv, bool GUIenabled)`](http://doc.qt.io/qt-4.8/qapplication.html#QApplication-2).
* The signature which allows to pass `QApplication::GuiServer` as third argument is: [`QApplication::QApplication(int & argc, char ** argv, Type type)`](http://doc.qt.io/qt-4.8/qapplication.html#QApplication-3).
* Both signature clashes: the one used in `wkhtmltopdf.cc` [forces the type to be either `QApplication::GuiClient` or `QApplication::Tty`](https://github.com/wkhtmltopdf/qt/blob/5a2c79e276cb880869e898c05a42d8949e1a8ec7/src/gui/kernel/qapplication.cpp#L769-L771).

Finally: you should pass `QApplication::GuiServer` has the third parameter passed to `QApplication` in [`wkhtmltopdf.cc`](https://github.com/wkhtmltopdf/wkhtmltopdf/blob/333521a51bf53ba632f3281a73fd14091c3ab250/src/pdf/wkhtmltopdf.cc#L189).

Note that, as stated by the documentation, you might not be able to run two or more wkhtmltopdf processes at the same time because they might clash (not sure however, you should test).

As mentioned above `QT_FONTS_ARE_RESOURCES` might be interesting if you know exactly which font will be used. But it might be also a bit tricky to modify wkhtmltopdf to embed fonts as resources if you don't have any knowledge about Qt. [Citing the Qt documentation](http://doc.qt.io/qt-4.8/resources.html):

> Qt resource system is a mechanism for storing binary files in the application's executable. This is useful if your application always needs a certain set of files

More generally, some debug techniques you could use to find the root cause of this issue (or of other issues):
* Use strace to see what's the last syscall right before it crashes
* Generate coredump when wkhtmltopdf segfaults, and open them with GDB

**Comment by ashkulz** [comment]
Please open a new issue in [wkhtmltopdf/packaging](https://github.com/wkhtmltopdf/packaging) if you want to continue this discussion.

**Comment by fanc-chen** [comment]
> Hi, I have transplanted wkthmltox to Mips platform. It can run but have some problems while I convert html to pdf. Follow is the detail.
> (note: I never learn C++)
> Step 1: I spent server days to transplant it to Mips platforms and it worked.
> Step 2: I upload it to embed-Mips device, run the command "./wkhtmltopdf test.html test.pdf". Here is the error msg:
> "QFontDatabase::loadFromCache: Could not open font database cache!""
> Step 3: I modified the source code: qt/src/gui/text/qfontdatabase_qws.cpp at line 164:
> `163 #ifdef Q_WS_QWS 164 // const bool weAreTheServer = QWSServer::instance(); // this is the orignal code 165 const bool weAreTheServer = true; // this is my modification 166 #else 167 const bool weAreTheServer = true; // assume single-process 168 #endif`
> Now, It can convert the pure English html files to pdf, but it core dumped if there have some Chinese chars in the html files.
> I think this is caused by my violent modification.
> May be it can worked well if "QWSServer::instance()" is true when we start the Qt, but I don't understand the C++ code. Please help me.
> Note: Some boy said it should add param '-aws' while you start the Qt.
> Thanks.

hello, wkhtmltopdf can run in mips platform？ How big is it after compilation？
